test/ranges.jl

```diff
@@ -1,6 +1,188 @@
+    return wu <= zu <= wu + lastbit
+end
+
+asbits(x) = reinterpret(Base.uinttype(typeof(x)), x)
+
+function roundshift(x, n)
+    xu = asbits(x)
+    lastbit = false
+    for i = 1:n
+        lastbit = isodd(xu)
+        xu = xu >> 1
+    end
+    xu + lastbit
+end
+
+subnormalmin(::Type{T}) where T = reinterpret(T, Base.uinttype(T)(1))
+
+function highprec_pair(x, y)
+    slopbits = (Base.Math.significand_bits(typeof(widen(x))) + 1) -
+        2*(Base.Math.significand_bits(typeof(x)) + 1)
+    hi, lo = Base.add12(x, y)
+    @test cmp_sn(widen(x) + widen(y), hi, lo)
+    hi, lo = Base.mul12(x, y)
+    @test cmp_sn(widen(x) * widen(y), hi, lo)
+    y == 0 && return nothing
+    hi, lo = Base.div12(x, y)
+    @test cmp_sn(widen(x) / widen(y), hi, lo, slopbits)
+    nothing
+end
+
+# # This tests every possible pair of Float16s. It takes too long for
+# # ordinary use, which is why it's commented out.
+# function pair16()
+#     for yu in 0x0000:0xffff
+#         for xu in 0x0000:0xffff
+#             x, y = reinterpret(Float16, xu), reinterpret(Float16, yu)
+#             highprec_pair(x, y)
+#         end
+#     end
+# end
+
+for T in (Float16, Float32) # skip Float64 (bit representation of BigFloat is not available)
+    for i = 1:10^5
+        x, y = rand(T), rand(T)
+        highprec_pair(x, y)
+        highprec_pair(-x, y)
+        highprec_pair(x, -y)
+        highprec_pair(-x, -y)
+    end
+    # Make sure we test dynamic range too
+    for i = 1:10^5
+        x, y = rand(T), rand(T)
+        x == 0 || y == 0 && continue
+        x, y = log(x), log(y)
+        highprec_pair(x, y)
+    end
+end
+
+asww(x) = widen(widen(x.hi)) + widen(widen(x.lo))
+astuple(x) = (x.hi, x.lo)
+
+function cmp_sn2(w, hi, lo, slopbits=0)
+    if !isfinite(hi)
+        if abs(w) > realmax(typeof(hi))
+            return isinf(hi) && sign(w) == sign(hi)
+        end
+        if isnan(w) && isnan(hi)
+            return true
+        end
+        return w == hi
+    end
+    if abs(w) < subnormalmin(typeof(hi))
+        return (hi == zero(hi) || abs(w - widen(hi)) < abs(w)) && lo == zero(hi)
+    end
+    z = widen(hi) + widen(lo)
+    zu, wu = asbits(z), asbits(w)
+    while zu > 0 && !isodd(zu)
+        zu = zu >> 1
+        wu = wu >> 1
+    end
+    zu = zu >> slopbits
+    wu = wu >> slopbits
+    return wu - 1 <= zu <= wu + 1
+end
+
+# TwicePrecision test. These routines lose accuracy if you form
+# intermediate subnormals; with Float16, this happens so frequently,
+# let's only test Float32.
+T = Float32
+Tw = widen(T)
+slopbits = (Base.Math.significand_bits(Tw) + 1) -
+    2*(Base.Math.significand_bits(T) + 1)
+for i = 1:10^5
+    x = Base.TwicePrecision{T}(rand())
+    y = Base.TwicePrecision{T}(rand())
+    xw, yw = asww(x), asww(y)
+    @test cmp_sn2(Tw(xw+yw), astuple(x+y)..., slopbits)
+    @test cmp_sn2(Tw(xw-yw), astuple(x-y)..., slopbits)
+    @test cmp_sn2(Tw(xw*yw), astuple(x*y)..., slopbits)
+    @test cmp_sn2(Tw(xw/yw), astuple(x/y)..., slopbits)
```

Comment:
Never, never any bugs, right? ;)  Sorry to be the bearer of bad news, but I imagine you'll be interested in https://github.com/JuliaLang/julia/issues/23497.

Reply:
By definition, if you think you've seen a bug in this code, you're just imagining things. Really. Honest. No bugs.

Ahhh! <sploosh>